text/0000-finalize-impl-dyn-syntax.md

```diff
@@ -0,0 +1,202 @@
+it needs to be done like this `(impl A + B)`, and not `impl (A + B)`. The second
+variant is a parsing error, but some people find it surprising and expect it to
+work, as if `impl` were an unary operator.
+
+## Alternative 2: low priority `+`
+
+Basically, `impl A + B` is parsed using same rules as `A + B`.
+
+If `impl A + B` is located inside a higher priority operator like `&` it has
+to be parenthesized.  
+If it is located at intersection of type and expressions
+grammars like `expr1 as Type + expr2`, it has to be parenthesized as well.
+
+`&dyn A + B` / `Fn() -> impl A + B` / `x as &dyn A + y` has to be rewritten as
+`&(dyn A + B)` / `Fn() -> (impl A + B)` / `x as &(dyn A + y)` respectively.
+
+One location must be mentioned specially, the location in a function return
+type:
+```rust
+fn f() -> impl A + B
+{
+    // Do things
+}
+```
+This is probably the most common location for `impl Trait` types.
+In theory, it doesn't require parentheses in any way - it's not inside of an
+unary operator and it doesn't cross expression boundaries.  
+However, it creates a bit of percieved inconsistency with function-like traits
+and function pointers that do require parentheses for `impl Trait` in return
+types (`Fn() -> (impl A + B)` / `fn() -> (impl A + B)`) because they, in their
+turn, can appear inside of unary operators and casts.  
+So, if avoiding this is considered more important than ergonomics, then
+we can require parentheses in function definitions as well.
+```rust
+fn f() -> (impl A + B)
+{
+    // Do things
+}
+```
+
+Pros:
+- Consistent priorities of binary and unary operators.
+- Parentheses are required relatively rarely (unless we require them in
+function definitions as well).
+
+Cons:
+- More parentheses than in the "Alternative 1".
+- `impl` / `dyn` is still a somewhat weird prefix construction and `dyn (A + B)`
+is not a valid syntax.
+
+## Alternative 3: Unary operator
+
+`impl` and `dyn` can become usual unary operators in type grammar like `&` or
+`*const`.  
+Their application to any other types except for (possibly parenthesized) paths
+(single `A`) or "legacy trait objects" (`A + B`) becomes an error, but this
+could be changed in the future if some other use is found.
+
+`&dyn A + B` / `Fn() -> impl A + B` / `x as &dyn A + y` has to be rewritten as
+`&dyn(A + B)` / `Fn() -> impl(A + B)` / `x as &dyn(A + y)` respectively.
+
+Function definitions with `impl A + B` in return type have to be rewritten too.
+```rust
+fn f() -> impl(A + B)
+{
+    // Do things
+}
+```
+
+Pros:
+- Consistent priorities of binary and unary operators.
+- `impl` / `dyn` are usual unary operators, `dyn (A + B)` is a valid syntax.
+
+Cons:
+- The largest amount of parentheses, parentheses are always required.  
+Parentheses are noise, there may be even less desire to use `dyn` in trait
+objects now, if something like `Box<Write + Send>` turns into
+`Box<dyn(Write + Send)>`.
```

Comment:
Would the space between `dyn` and the open parenthesis be required by the grammar?

Reply:
@stuhood 
> The awkwardness of that case would seem to stem more from the fact that Box already has implicit behaviour that causes a vtable to be created (ie, that dyn is implicit with Box currently), and less from the particular syntax that is in use.

I think maybe you misunderstood my comment-- it's primarily `Box<dyn<Write>>` that I dislike more than `Box<dyn Write>`-- `Box<dyn<Write + Read>>` and `Box<dyn(Write + Read)>` are very similar except for the fact that parentheses are what I would try first when encountering a precedence error.